src/Dependencies/Collections/SegmentedArray`1.cs

```diff
@@ -71,7 +70,7 @@ public SegmentedArray(int length)
                 // Avoid using (length & s_offsetMask) because it doesn't handle a last page size of s_segmentSize.
                 var lastPageSize = length - ((_items.Length - 1) << s_segmentShift);
 
-                _items[^1] = new T[lastPageSize];
+                _items[_items.Length - 1] = new T[lastPageSize];
```

Comment:
What's the purpose of this change? Does it just not have the range types available?

Reply:
Correct. It's easy to assume the files are provided in dotnet/roslyn, but we can't make the same assumption if the shared project is consumed by another repository.

Reply:
> but we can't make the same assumption if the shared project is consumed by another repository.

Is that a goal of this change? If so, why?

Reply:
I do make the assumption that **NullableAnnotations.cs** is provided. This seems to be a common expectation.

Reply:
What about the more general question though? Who is going to be consuming _source code_ from the dotnet/roslyn repository, and why is that a goal for this change? It seems to me that that would be an anti-goal: it's like an IVT, but without even the advertisement of an `InternalsVisibleTo` attribute.

Reply:
Current candidates are dotnet/project-system, CPS, and MSBuild.

Reply:
We already use source packages to share Debugging helpers and PooledObjects. Source packages do not have the same issues as IVTs. We can make any changes, even breaking without breaking anyone. This is because the project that uses the source package ends up having their own copy of the source that won't get updated unless the project explicitly does so and builds in their CI. 

---
In reply to: [543702405](https://github.com/dotnet/roslyn/pull/49980#discussion_r543702405) [](ancestors = 543702405)